Orange/widgets/data/owmergedata.py

```diff
@@ -362,20 +364,34 @@ def _join_table_by_indices(self, reduced_extra, indices):
     def _join_array_by_indices(left, right, indices, string_cols=None):
         """Join (horizontally) two arrays, taking pairs of rows given in indices
         """
+        def matrix(height, width, dtype, sparse=False):
+            if not sparse:
+                return np.full((height, width), np.nan, dtype=dtype)
+            else:
+                return sp.csr_matrix((height, width), dtype=dtype)
+
         tpe = object if object in (left.dtype, right.dtype) else left.dtype
         left_width, right_width = left.shape[1], right.shape[1]
-        arr = np.full((indices.shape[1], left_width + right_width), np.nan, tpe)
+
+        sparse = sp.issparse(left) or sp.issparse(right)
+
+        arr1 = matrix(indices.shape[1], left_width, tpe, sparse)
+        arr2 = matrix(indices.shape[1], right_width, tpe, sparse)
         if string_cols:
-            arr[:, string_cols] = ""
-        for indices, to_change, lookup in (
-                (indices[0], arr[:, :left_width], left),
-                (indices[1], arr[:, left_width:], right)):
-            known = indices != -1
-            to_change[known] = lookup[indices[known]]
+            arr1[:, [sc for sc in string_cols if sc < left_width]] = ""
+            arr2[:, [sc - left_width for sc in string_cols if sc >= left_width]] = ""
+
+        for ind, to_change, lookup in (
+                (indices[0], arr1, left),
+                (indices[1], arr2, right)):
+            known = ind != -1
+            if sum(known):
+                to_change[known] = lookup[ind[known]]
```

Comment:
This is probably not sufficient for sparse data. For dense, we initially set all values as missing and hence it's ok to set only known values. For sparse, we initially set all values as zeros. Hence we should, along setting known values, also set unknown values as nans.